File: workflow/src/lib/extract-form-data.ts
import * as cheerio from "cheerio";

export interface FormSection {
  sectionTitle: string | null;
  options: string[];
}

export interface FormData {
  title: string;
  sections: FormSection[];
}

/**
 * Extracts form data from Google Form HTML.
 * Tries FB_PUBLIC_LOAD_DATA_ JSON blob first, falls back to HTML parsing.
 * Filters to only date/time sections.
 */
export function extractFormData(html: string): FormData {
  const loadDataResult = parseFromLoadData(html);
  const raw = loadDataResult ?? parseFromHtml(html);

  return {
    title: raw.title,
    sections: filterDateSections(raw.sections),
  };
}

/**
 * Keeps only sections where at least some options look like dates/times.
 * Also strips empty strings and "Not Available" entries from options.
 */
export function filterDateSections(sections: FormSection[]): FormSection[] {
  return sections
    .filter((s) => s.options.some(isDateTimeOption))
    .map((s) => ({
      ...s,
      options: s.options.filter((o) => o.trim() !== "" && isDateTimeOption(o)),
    }));
}

const DATE_TIME_PATTERN =
  /\b(mon|tue|wed|thu|fri|sat|sun|january|february|march|april|may|june|july|august|september|october|november|december|jan|feb|mar|apr|jun|jul|aug|sep|oct|nov|dec|\d{1,2}\/\d{1,2}|\d{1,2}:\d{2}\s*(am|pm)?)\b/i;

function isDateTimeOption(text: string): boolean {
  return DATE_TIME_PATTERN.test(text);
}

/**
 * Parses Google Form data from the FB_PUBLIC_LOAD_DATA_ JavaScript variable.
 *
 * Google Forms embed all form structure in a JS variable like:
 *   var FB_PUBLIC_LOAD_DATA_ = [...deeply nested arrays...];
 *
 * The structure (as of 2025):
 * - [1][1] = form title
 * - [1][1] or root level arrays contain question groups
 * - Checkbox questions have type indicator 4
 * - Each option is a nested array with the label at index [0]
 */
export function parseFromLoadData(html: string): FormData | null {
  const match = html.match(
    /var\s+FB_PUBLIC_LOAD_DATA_\s*=\s*([\s\S]*?);\s*<\/script>/
  );
  if (!match) return null;

  let data: unknown;
  try {
    data = JSON.parse(match[1]);
  } catch {
    return null;
  }

  if (!Array.isArray(data)) return null;

  // data[1][8] is the form title in most observed structures
  // data[1][1] is an array of question groups
  const formInfo = (data as unknown[])[1];
  if (!Array.isArray(formInfo)) return null;

  const title = typeof formInfo[8] === "string" ? formInfo[8] : "Google Form";

  // formInfo[1] contains the list of form items/questions
  const items = formInfo[1];
  if (!Array.isArray(items)) return { title, sections: [] };

  const sections: FormSection[] = [];

  for (const item of items) {
    if (!Array.isArray(item)) continue;

    // item[1] = question title
    // item[3] = question type (4 = checkboxes)
    // item[4] = array of question parts containing options
    const sectionTitle =
      typeof item[1] === "string" ? item[1] : null;
    const questionType = item[3];
    const questionParts = item[4];

    if (questionType !== 4 || !Array.isArray(questionParts)) continue;

    for (const part of questionParts) {
      if (!Array.isArray(part)) continue;

      // part[1] = array of options
      const optionsArray = part[1];
      if (!Array.isArray(optionsArray)) continue;

      const options: string[] = [];
      for (const opt of optionsArray) {
        if (Array.isArray(opt) && typeof opt[0] === "string") {
          options.push(opt[0]);
        }
      }

      if (options.length > 0) {
        sections.push({ sectionTitle, options });
      }
    }
  }

  return { title, sections };
}

/**
 * Fallback: parses Google Form HTML using cheerio to find checkbox options.
 */
export function parseFromHtml(html: string): FormData {
  const $ = cheerio.load(html);

  // Try to get the form title
  const title =
    $('meta[property="og:title"]').attr("content") ||
    $("title").text().replace(/ - Google Forms$/, "") ||
    "Google Form";

  const sections: FormSection[] = [];

  // Google Forms renders checkbox groups in div containers
  // Each group has a heading and a list of options with data-value or label text
  $('[role="group"], [role="listbox"], .freebirdFormviewerComponentsQuestionCheckboxRoot').each(
    (_i, group) => {
      const $group = $(group);
      const sectionTitle =
        $group
          .closest('[data-params]')
          .find('[role="heading"]')
          .text()
          .trim() || null;

      const options: string[] = [];
      $group.find('[data-answer-value], [data-value]').each((_j, opt) => {
        const val =
          $(opt).attr("data-answer-value") || $(opt).attr("data-value");
        if (val) options.push(val);
      });

      // Fallback: look for label text
      if (options.length === 0) {
        $group.find('label, [role="option"]').each((_j, opt) => {
          const text = $(opt).text().trim();
          if (text) options.push(text);
        });
      }

      if (options.length > 0) {
        sections.push({ sectionTitle, options });
      }
    }
  );

  return { title, sections };
}

File: modal/lib/extract_form_data.py
import json
import re
from dataclasses import dataclass, field
from typing import Optional

from bs4 import BeautifulSoup


@dataclass
class FormSection:
    section_title: Optional[str]
    options: list[str]


@dataclass
class FormData:
    title: str
    sections: list[FormSection] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "title": self.title,
            "sections": [
                {
                    "sectionTitle": s.section_title,
                    "options": s.options,
                }
                for s in self.sections
            ],
        }


DATE_TIME_PATTERN = re.compile(
    r"\b(mon|tue|wed|thu|fri|sat|sun"
    r"|january|february|march|april|may|june|july|august|september|october|november|december"
    r"|jan|feb|mar|apr|jun|jul|aug|sep|oct|nov|dec"
    r"|\d{1,2}/\d{1,2}"
    r"|\d{1,2}:\d{2}\s*(?:am|pm)?)\b",
    re.IGNORECASE,
)


def _is_date_time_option(text: str) -> bool:
    return bool(DATE_TIME_PATTERN.search(text))


def filter_date_sections(sections: list[FormSection]) -> list[FormSection]:
    result = []
    for s in sections:
        if not any(_is_date_time_option(o) for o in s.options):
            continue
        filtered_options = [
            o for o in s.options if o.strip() != "" and _is_date_time_option(o)
        ]
        result.append(FormSection(section_title=s.section_title, options=filtered_options))
    return result


def parse_from_load_data(html: str) -> Optional[FormData]:
    match = re.search(
        r"var\s+FB_PUBLIC_LOAD_DATA_\s*=\s*([\s\S]*?);\s*</script>", html
    )
    if not match:
        return None

    try:
        data = json.loads(match.group(1))
    except (json.JSONDecodeError, ValueError):
        return None

    if not isinstance(data, list):
        return None

    form_info = data[1] if len(data) > 1 else None
    if not isinstance(form_info, list):
        return None

    title = form_info[8] if len(form_info) > 8 and isinstance(form_info[8], str) else "Google Form"

    items = form_info[1] if len(form_info) > 1 else None
    if not isinstance(items, list):
        return FormData(title=title)

    sections: list[FormSection] = []

    for item in items:
        if not isinstance(item, list):
            continue

        section_title = item[1] if len(item) > 1 and isinstance(item[1], str) else None
        question_type = item[3] if len(item) > 3 else None
        question_parts = item[4] if len(item) > 4 else None

        if question_type != 4 or not isinstance(question_parts, list):
            continue

        for part in question_parts:
            if not isinstance(part, list):
                continue

            options_array = part[1] if len(part) > 1 else None
            if not isinstance(options_array, list):
                continue

            options: list[str] = []
            for opt in options_array:
                if isinstance(opt, list) and len(opt) > 0 and isinstance(opt[0], str):
                    options.append(opt[0])

            if options:
                sections.append(FormSection(section_title=section_title, options=options))

    return FormData(title=title, sections=sections)


def parse_from_html(html: str) -> FormData:
    soup = BeautifulSoup(html, "html.parser")

    og_title = soup.find("meta", property="og:title")
    if og_title and og_title.get("content"):
        title = og_title["content"]
    elif soup.title and soup.title.string:
        title = re.sub(r"\s*-\s*Google Forms$", "", soup.title.string)
    else:
        title = "Google Form"

    sections: list[FormSection] = []

    for group in soup.find_all(attrs={"role": ["group", "listbox"]}):
        data_params_parent = group.find_parent(attrs={"data-params": True})
        heading_el = data_params_parent.find(attrs={"role": "heading"}) if data_params_parent else None
        section_title = heading_el.get_text(strip=True) if heading_el else None

        options: list[str] = []
        for opt in group.find_all(attrs={"data-answer-value": True}):
            val = opt.get("data-answer-value")
            if val:
                options.append(val)

        if not options:
            for opt in group.find_all(attrs={"data-value": True}):
                val = opt.get("data-value")
                if val:
                    options.append(val)

        if not options:
            for opt in group.find_all(["label", lambda tag: tag.get("role") == "option"]):
                text = opt.get_text(strip=True)
                if text:
                    options.append(text)

        if options:
            sections.append(FormSection(section_title=section_title, options=options))

    return FormData(title=title, sections=sections)


def extract_form_data(html: str) -> FormData:
    load_data_result = parse_from_load_data(html)
    raw = load_data_result if load_data_result is not None else parse_from_html(html)

    return FormData(
        title=raw.title,
        sections=filter_date_sections(raw.sections),
    )
